src/main/java/io/cdap/cdap/metrics/process/gcp/GoogleCloudMonitoringWriter.java

```diff
@@ -0,0 +1,219 @@
+
+import com.google.cloud.monitoring.v3.MetricServiceClient;
+import com.google.cloud.monitoring.v3.MetricServiceSettings;
+import com.google.gson.Gson;
+import com.google.monitoring.v3.TimeSeries;
+import io.cdap.cdap.api.metrics.MetricType;
+import io.cdap.cdap.api.metrics.MetricValue;
+import io.cdap.cdap.api.metrics.MetricValues;
+import io.cdap.cdap.api.metrics.MetricsWriter;
+import io.cdap.cdap.api.metrics.MetricsWriterContext;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.FileReader;
+import java.io.Reader;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.concurrent.atomic.AtomicLong;
+import java.util.stream.Collectors;
+import javax.annotation.Nullable;
+
+/**
+ * {@link MetricsWriter} implementation for writing metrics to Google Cloud Monitoring
+ */
+public class GoogleCloudMonitoringWriter implements MetricsWriter {
+
+  private static final Logger LOG = LoggerFactory.getLogger(GoogleCloudMonitoringWriter.class);
+  private static final Gson GSON = new Gson();
+  private static final String RESOURCE_CONTAINER = "resource_container";
+  private static final String PROJECT = "project";
+  private static final String ORG_ID = "org_id";
+  private static final String LOCATION = "location";
+  private static final String ANTHOS_CLUSTER = "anthos_cluster";
+  private static final String INSTANCE_ID = "instance_id";
+  private static final String VERSION = "version";
+  private static final String POLL_FREQUENCY_SECONDS = "poll.frequency.seconds";
+  private static final String MONITORING_ENDPOINT = "monitoring.endpoint";
+  private static final String CONFIG_FILE_PATH = "config.file.path";
+
+  private final AtomicLong lastEndTime = new AtomicLong(-1);
+  private Map<String, MetricsMapping> metricsMapping;
+  @Nullable
+  private MetricServiceClient metricServiceClient;
+  @Nullable
+  private String projectName;
+  private Map<String, String> autoFilledLabelMap;
+  private int pollFreqInSeconds;
+
+  @Override
+  public void write(Collection<MetricValues> metricValues) {
+    if (metricServiceClient == null) {
+      LOG.debug(
+        "MetricServiceClient is not available. Returning with out writing values in GoogleCloudMonitoringWriter.");
+      return;
+    }
+    if (projectName == null) {
+      LOG.debug(
+        "Project name is not configured. Returning with out writing values in GoogleCloudMonitoringWriter.");
+      return;
+    }
+    if (metricsMapping.isEmpty()) {
+      LOG.debug(
+        "Metrics are not mapped. Returning with out writing values in GoogleCloudMonitoringWriter.");
+      return;
+    }
+    LOG.debug("Writing metric values in GoogleCloudMonitoringWriter");
+    try {
+      Map<TimeSeriesMetadata, List<Long>> timeSeriesMap = createTimeSeriesMap(metricValues);
+      long endTime = System.currentTimeMillis() / 1000;
```

Comment:
nit: name these as endTimeMilliSecond, so the unit is clear

Reply:
Done.